java/compaction-core/src/main/java/sleeper/compaction/job/status/CompactionJobStatusesBuilder.java

```diff
@@ -15,60 +15,98 @@
     }
 
-    public CompactionJobStatusesBuilder expiryDate(
-            String jobId, Instant expiryDate) {
-        expiryDateById.put(jobId, expiryDate);
-        return this;
-    }
-
     public List<CompactionJobStatus> build() {
         return stream().collect(Collectors.toList());
     }
 
-    private CompactionJobStatus fullStatus(String jobId, CompactionJobCreatedStatus created) {
+    private CompactionJobStatus fullStatus(String jobId, CompactionJobStatusUpdateRecord createdUpdate) {
+        CompactionJobCreatedStatus createdStatus = (CompactionJobCreatedStatus) createdUpdate.getStatusUpdate();
+        List<CompactionJobStatusUpdateRecord> runUpdates = runUpdatesOrderedByUpdateTime(jobId);
+
         return CompactionJobStatus.builder().jobId(jobId)
-                .createdStatus(created)
-                .startedStatus(startedById.get(jobId))
-                .finishedStatus(finishedById.get(jobId))
-                .expiryDate(expiryDateById.get(jobId))
+                .createdStatus(createdStatus)
+                .jobRuns(buildJobRunList(runUpdates))
+                .expiryDate(last(runUpdates)
+                        .map(CompactionJobStatusUpdateRecord::getExpiryDate)
+                        .orElseGet(createdUpdate::getExpiryDate))
                 .build();
     }
+
+    public List<CompactionJobRun> buildJobRunList(List<CompactionJobStatusUpdateRecord> recordList) {
+        Map<String, CompactionJobRun.Builder> taskBuilders = new HashMap<>();
+        List<CompactionJobRun> jobRuns = new ArrayList<>();
+        for (CompactionJobStatusUpdateRecord record : recordList) {
+            String taskId = record.getTaskId();
+            CompactionJobStatusUpdate statusUpdate = record.getStatusUpdate();
+            if (statusUpdate instanceof CompactionJobStartedStatus) {
+                if (taskBuilders.containsKey(taskId)) {
+                    jobRuns.add(taskBuilders.remove(taskId).build());
+                }
+                taskBuilders.computeIfAbsent(taskId, id -> CompactionJobRun.builder())
+                        .startedStatus((CompactionJobStartedStatus) statusUpdate)
+                        .taskId(taskId);
+            } else {
+                jobRuns.add(taskBuilders.remove(taskId)
+                        .finishedStatus((CompactionJobFinishedStatus) statusUpdate)
+                        .taskId(taskId).build());
+            }
+        }
+        taskBuilders.values().stream().map(CompactionJobRun.Builder::build).forEach(jobRuns::add);
+        return jobRuns;
+    }
+
+    private Map<String, CompactionJobStatusUpdateRecord> createUpdatesOrderedByUpdateTime() {
+        return createdUpdateByJobId.entrySet()
+                .stream().sorted(Comparator.comparing(update -> update.getValue().getStatusUpdate().getUpdateTime()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (oldValue, newValue) -> oldValue, TreeMap::new));
```

Comment:
I don't think this is needed? The sorted call won't do anything since the TreeMap will sort everything but the natural order of the key, ie. the job ID.